corehq/motech/openmrs/handler.py

```diff
@@ -72,125 +82,139 @@ def run(self, requests, info, openmrs_config, person_uuid, attributes):
-                        requests, person_uuid, attribute_uuid, person_attribute_type, existing_value,
-                    )
-                    self._subtasks.append(
+                    subtasks.append(
                         WorkflowTask(
-                            rollback_task,
-                            None,  # pass_result_as=None because both tasks are updates; rollback just resets
-                            update_person_attribute,
-                            requests, person_uuid, attribute_uuid, person_attribute_type, value,
+                            func=update_person_attribute,
+                            func_args=(requests, person_uuid, attribute_uuid, person_attribute_type, value),
+                            rollback_func=update_person_attribute,
+                            rollback_args=(requests, person_uuid, attribute_uuid, person_attribute_type,
+                                           existing_value),
                         )
                     )
             else:
-                self._subtasks.append(
+                subtasks.append(
                     WorkflowTask(
-                        delete_person_attribute_task(requests, person_uuid),
-                        'attribute_uuid',  # create_person_attribute returns attribute_uuid: Pass to rollback task.
-
-                        # Note: We can't use the `@workflow_task` decorator on create_person_attribute() because we
-                        # need to call delete_person_attribute_task() with values for `requests` and `person_uuid`
-                        # for it to return a Task instance for WorkflowTask.rollback_task.
-                        create_person_attribute,
-                        requests, person_uuid, person_attribute_type, value,
+                        func=create_person_attribute,
+                        func_args=(requests, person_uuid, person_attribute_type, value),
+                        rollback_func=delete_person_attribute,
+                        rollback_args=(requests, person_uuid),
+                        pass_result=True
                     )
                 )
+        return subtasks
 
 
 class CreateVisitsTask(WorkflowTask):
 
     def __init__(self, *args, **kwargs):
-        super(CreateVisitsTask, self).__init__(None, None, None, *args, **kwargs)
+        super(CreateVisitsTask, self).__init__(func=None, func_args=args, func_kwargs=kwargs)
 
-    def run(self, requests, info, form_json, form_question_values, openmrs_config, person_uuid):
+    def run(self):
         """
         This task does not make any changes itself. It uses subtasks for creating visits, encounters and
         observations
         """
+        requests, info, form_json, form_question_values, openmrs_config, person_uuid = unpack_args(
+            ('requests', 'info', 'form_json', 'form_question_values', 'openmrs_config', 'person_uuid'),
+            *self.func_args, **self.func_kwargs,
+        )
```

Comment:
I am not loving this construct. It reminds me of bad Perl. Any ideas @snopoke ? The alternative is not to override the same method that is called by `execute_workflow()` but rather have something like ...
```python
    def __run(self, requests, info, form_json, form_question_values, openmrs_config, person_uuid):
        # ...
```
... that `run()` calls if `self.func` is None.